### PR TITLE
[Android] fixes Picker list showing up incorrectly when focus is set on other controls

### LINKED ISSUE #4187: [Android] Picker list shows up, when focus is set on other controls.
### Description

Android issue. 
Xamarin.Form 3.3.0.912540 issue. Works ok in 3.2.0.871581

When setting the focus on some other control, picker list shows up. when picker and other controls are inside ListVew.DataTemplate.ViewCell.
[App2.zip](https://github.com/xamarin/Xamarin.Forms/files/2508026/App2.zip)

### Steps to Reproduce

1. Run the attached sample solution
2. Select a value from Picker
3. It works correctly
4. Set focus on Entry
5. The picker list shows up
6. That is incorrect 

### Expected Behavior

when setting the focus on the entry, picker list should not show up.

### Actual Behavior

when setting the focus on the entry, picker list shows up.

### Basic Information

- Versions with issue: Xamarin.Form 3.3.0.912540, 3.3.0.967583, 3.4.0.1008975
- Last known good version: Xamarin.Form 3.2.0.871581
- IDE: 15.8.7
- Platform Target Frameworks: 
  - Android: 9 

### Sample

[App2.zip](https://github.com/xamarin/Xamarin.Forms/files/2508028/App2.zip)

### LINKED ISSUE #4318: [Android] Pop Up a picker when changing the page of CarouselPage
### Description

When changed page of `CarouselPage` pop-ups `Picker` control as if he was pressed.

### Steps to Reproduce

1.  Create CarouselPage 
2.  Add 2 pages with Picker control
3.  Focus on a `Picker`
4.  Change page

### Expected Behavior

Picker don't pops up

### Actual Behavior

Picker pops up

### Basic Information

- Version with issue: master branch
- Last known good version: -
- IDE: Windows Visual Studio 2017 15.8.8

### Screencast

http://recordit.co/FkHj45S1pK

### Reproduction Link

https://github.com/paymicro/Xamarin.Forms/commit/1d1d01a89e0b78fd47010ad18b7ce346cca6897b

### LINKED ISSUE #4494: DatePicker inside TableView raises popup dialogs unexpectedly
### Description
When used inside TableView's ViewCell, a DatePicker will sometimes raise its pop up dialog unexpectedly after its value is changed.

### Steps to Reproduce

1. Run the provided sample app in the Android emulator.
2. Change the start date to 11/26.
3. Change the end date to 11/29.

### Expected Behavior
No further dialogs pop up after user activity has ceased.

### Actual Behavior
After setting the end date, two DatePicker dialogs pop up again without user intervention.

### Basic Information

- Version with issue: 3.4.0.1008975
- Last known good version: Unknown
- IDE: VS 2017 15.9.2
- Platform Target Frameworks: 
  - Android: 8.1
- Android Support Library Version: 27.0.2.1
- Nuget Packages: None
- Affected Devices: N/A

### Reproduction Link

[DatePicker.zip](https://github.com/xamarin/Xamarin.Forms/files/2602868/DatePicker.zip)

### PR DESCRIPTION
**Pull request by paymicro**
### Description of Change ###

The picker dialog is shown only if the focus was obtained by tabulation or touch event.

### Issues Resolved ### 

- fixes #4318
- fixes #4187
- fixes #4494

### API Changes ###
 
 None

### Platforms Affected ### 

- Android

### Before/After Screenshots ### 

Before
http://recordit.co/IULOwzWY2k
After
http://recordit.co/GjiYmiSFQv

### Testing Procedure ###
**In cell**
- run UItest 4187
- single click on the Picker and select any option
- click on any TextBox (the dialogue with the options of the picker should't open)

**Carousel page**
- run UItest 4187
- single click on the upper Picker and select any option
- swipe to next page (the dialogue with the options of the picker should't open)

**Tab stops**
- run UItest 1700
- single click on the Picker and select any option
- check that the Picker's dialogue opens if he gets the focus at the tabulation

### PR Checklist ###

- [ ] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Platform.Android/Renderers/PickerRenderer.cs [hunk 1]
```diff
@@ -1,33 +1,22 @@
...
 using Orientation = Android.Widget.Orientation;
 using Android.Content;
-using Xamarin.Forms.PlatformConfiguration.AndroidSpecific;
-using System.Collections.Generic;
-using Android.Text;
 
 namespace Xamarin.Forms.Platform.Android
 {
-	public class PickerRenderer : ViewRenderer<Picker, EditText>
+	public class PickerRenderer : ViewRenderer<Picker, PickerEditText>, IPickerRenderer
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.Android/Renderers/PickerRenderer.cs]: I think we can leave EditText

File: Xamarin.Forms.Platform.Android/Renderers/PickerEditText.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		{
+			Focusable = true;
+			Clickable = true;
+			InputType = InputTypes.Null;
+			KeyPress += OnKeyPress;
+			owner = pickerRenderer;
+			SetOnClickListener(PickerListener.Instance);
+		}
+
+		private void OnKeyPress(object sender, KeyEventArgs e)
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.Android/Renderers/PickerEditText.cs]: drop private

File: Xamarin.Forms.Platform.Android/Renderers/PickerEditText.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Xamarin.Forms.Platform.Android
+{
+	public class PickerEditText : EditText
+	{
+		HashSet<Keycode> availableKeys = new HashSet<Keycode>(new[] {
+			Keycode.Tab, Keycode.Forward, Keycode.Back, Keycode.DpadDown, Keycode.DpadLeft, Keycode.DpadRight, Keycode.DpadUp
+		});
+
+		readonly IPickerRenderer owner;
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.Android/Renderers/PickerEditText.cs]: Let's turn this in a WeakReference

**Comment by rmarinho** [comment]
Small changes, but wires, fixes the issue and tab index still works. 
Good work

**Comment by malsamaxyz** [comment]
Is this related to this?:
https://stackoverflow.com/questions/53210901/xamarin-forms-picker-control-opens-when-entering-input-in-entry-control

**Comment by paymicro** [comment]
@malsamaxyz yes. This PR should fix this problem.

**Comment by rmarinho** [comment]
Can you add a UITEst ? that would be great. Something that shows that when focus is set doesn't always show the picker, for example for the carousel page change with a picker inside. 

Thanks

**Comment by rmarinho** [comment]
@paymicro failing 4187 test 
```
Carousel - #1
Expected: True
But was: False
```

**Comment by hartez** [comment]
When I run the UI test locally, it's getting stuck with the TimePicker dialog open. I think it's trying to tap on "Text 1", but "Text 1" is covered partially by the dialog so it's just tapping on the dialog and leaving it open.

Since the test only runs on Android, you can probably just dismiss the dialog with `RunningApp.Back()`.

**Comment by rmarinho** [comment]
@paymicro you need to check if you enable access from maintainers to your fork, that way I could push to your fork. 

I think we also fix issue #4494  with this fix. Can you add this commit there and double check with that repo to see if we are really covering the case. 

https://github.com/xamarin/Xamarin.Forms/commit/a7ad11d02708ffedad81d57e69de41b67ab342ab

Thanks

**Comment by samhouts** [comment]
@paymicro Unfortunately, this breaks the test in Bugzilla41424. Can you take a look at that and see if you can fix it? Thanks!

**Pull request merged**